### PR TITLE
Rewrite: libAnki Sound.kt and add TTS Support

### LINKED ISSUE #14358: Add support for TTS tags
Currently AnkiDroid is able to get the text from TTS tags from the backend, but the code is not currently set up to feed the text into the TTS engine. controlSound() needs to be updated to do that, and the `@BlocksSchemaUpgrade()` line should be removed. We'll also need to ensure the text plays correctly by default, and not just when a play icon is clicked on.

### PR DESCRIPTION
**Pull request by david-allison**
Disclaimer 💰: I was offered financial sponsorship for TTS Support, but don't know if this is still going ahead. Let me know how you'd like to handle this

## Purpose / Description
Our sound code had a lot of unnecessary complexity due to the usage of callbacks in the underlying Android APIs

Kotlin allows us to use coroutines instead, which significantly simplifies the calling process

## Fixes
* Fixes #14358
* Technically there's a small 'enhancement' follow-up to ensure that `ar` is compatible with Anki Desktop

## Approach
* Follow on from https://github.com/ankidroid/Anki-Android/pull/14967
* rewrite `Sound`
  * `SoundPlayer` coordinates `SoundTagPlayer` and `TtsPlayer`
  * Use task cancellation to stop tasks
  * Use coroutines to set up `SoundTagPlayer`
* Instead of a timeout to Automatic Display Answer, use a callback
* add tests
* Spend far too long fixing tests

## How Has This Been Tested?
Unit tested + tested on my S21
* Automatic Display Answer
* TTS
* Missing TTS Voice

Fixed a bug, but needs some thought

Play TTS, edit note, return. After which, TTS didn't work
* I believed this was caused by `stop()` not being handled
* but it turned out that the lifecycle owner also wasn't subscribed

image

## Learning (optional, can help others)
* Testing probably needs to be manual: cancellation is cancelling a `Job` rather than calling `.stop()` and this was difficult to simulate acceptably in Unit Tests

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by david-allison** [comment]
Also made this test flaky:

```
AbstractFlashcardViewerTest > noAutomaticAnswerAfterRenderProcessGoneAndPaused_issue9632 FAILED
    java.lang.AssertionError: messages after flipping card
    Expected: <true>
         but: was <false>
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
        at com.ichi2.anki.AbstractFlashcardViewerTest.noAutomaticAnswerAfterRenderProcessGoneAndPaused_issue9632(AbstractFlashcardViewerTest.kt:246)
```

**Comment by david-allison** [comment]
<details>

```
W/PackageParser: Unknown element under <manifest>: queries at /Users/davidallison/StudioProjects/Anki-Android/AnkiDroid/build/intermediates/apk_for_local_test/playDebugUnitTest/apk-for-local-test.ap_ Binary XML file line #59
W/PackageParser: No actions in intent filter at /Users/davidallison/StudioProjects/Anki-Android/AnkiDroid/build/intermediates/apk_for_local_test/playDebugUnitTest/apk-for-local-test.ap_ Binary XML file line #322
W/ACRA: ACRA#init called more than once. This might have unexpected side effects. Doing this outside of tests is discouraged.
I/ACRA: ACRA is enabled for com.ichi2.anki.debug, initializing...
D/AnkiDroid: Startup - Application Start
I/UsageAnalytics: initialize()
I/UsageAnalytics: setOptIn(): from false to false
I/GoogleAnalyticsImpl: Not participating in analytics sample (sample percentage vs random: 10 36)
D/UsageAnalytics: setOptIn() optIn / sAnalytics.config().enabled(): false/false
I/UsageAnalytics: setDryRun(): true, warning dryRun is experimental
D/AdaptionUtil: isRunningAsUnitTest: true
D/SystemContextMenu: Modifying Context Menu Status: Preference was true
I/NotificationChannels: Creating notification channel with id/name: General Notifications/AnkiDroid
I/NotificationChannels: Creating notification channel with id/name: Synchronization/Synchronization
I/NotificationChannels: Creating notification channel with id/name: Global Reminders/Cards due
I/NotificationChannels: Creating notification channel with id/name: Deck Reminders/Reminders
I/NotificationChannels: Creating notification channel with id/name: Scoped Storage/Storage migration
I/AnkiDroidApp: AnkiDroidApp: Starting Services
W/BootService: Boot Service did not execute - no permissions
D/TtsVoices: job already started
D/Backend: Opening rust backend with lang=[en]
I/Collection: (Re)opening Database: /var/folders/ym/nqynp93d4j74dpzw20sq4wq00000gn/T/robolectric-AbstractFlashcardViewerTest_noAutomaticAnswerAfterRenderProcessGoneAndPaused_issue96326369019436638851077/external-files/AnkiDroid/collection.anki2
I/Themes: Setting theme to LIGHT
I/AnkiDroidApp$onCreate: NonAbstractFlashcardViewer::onCreate
D/MotionEventHandler$Companion: setting up 0 motion handlers
Invalid ID 0x00000000.
I/ReviewerServer: Starting server on http://127.0.0.1:60501
I/AnkiDroidApp$onCreate: NonAbstractFlashcardViewer::onStart
I/ScopedStorageService: isLegacyStorage(): false
I/AnkiDroidApp$onCreate: NonAbstractFlashcardViewer::onResume
D/UsageAnalytics: sendAnalyticsScreenView(): NonAbstractFlashcardViewer
D/UsageAnalytics: getOptIn() status: false
D/TtsVoices: job already started
D/Note: load()
D/AbstractFlashcardViewer: displayCardQuestion()
D/AbstractFlashcardViewer: Focusable = true, Focusable in touch mode = false
D/AbstractFlashcardViewer$LinkDetectingGestureDetector: Initializing WebView touch handler
D/AbstractFlashcardViewer: updateCard()
I/SoundPlayer: loading sounds for card 1702910905548 (FRONT)
I/SoundPlayer: stopping sounds
V/CardSoundConfig$Companion: start loading SoundConfig
D/CardSoundConfig$Companion: loaded SoundConfig: com.ichi2.anki.cardviewer.CardSoundConfig@4d3933ab
V/CardHtml: ::style:: / BODY {font-family: 'OpenSans';font-weight: normal;font-style: normal;font-stretch: normal;}
D/AbstractFlashcardViewer: base url = file:///var/folders/ym/nqynp93d4j74dpzw20sq4wq00000gn/T/robolectric-AbstractFlashcardViewerTest_noAutomaticAnswerAfterRenderProcessGoneAndPaused_issue96326369019436638851077/external-files/AnkiDroid/collection.media/
D/AbstractFlashcardViewer: fillFlashcard()
D/AbstractFlashcardViewer: base url = file:///var/folders/ym/nqynp93d4j74dpzw20sq4wq00000gn/T/robolectric-AbstractFlashcardViewerTest_noAutomaticAnswerAfterRenderProcessGoneAndPaused_issue96326369019436638851077/external-files/AnkiDroid/collection.media/
D/AbstractFlashcardViewer$LinkDetectingGestureDetector: Removing pending touch events for gestures
I/AbstractFlashcardViewer: AbstractFlashcardViewer:: Question successfully shown for card id 1702910905548
I/SoundPlayer: playing sounds for QUESTION
V/AbstractFlashcardViewer: onSoundGroupCompleted
D/AbstractFlashcardViewer: displayCardAnswer()
D/TypeAnswer: correct answer = null
D/TypeAnswer: user answer = 
D/AbstractFlashcardViewer: updateCard()
I/SoundPlayer: loading sounds for card 1702910905548 (BACK)
I/SoundPlayer: stopping sounds
V/CardHtml: ::style:: / BODY {font-family: 'OpenSans';font-weight: normal;font-style: normal;font-stretch: normal;}
D/AbstractFlashcardViewer: base url = file:///var/folders/ym/nqynp93d4j74dpzw20sq4wq00000gn/T/robolectric-AbstractFlashcardViewerTest_noAutomaticAnswerAfterRenderProcessGoneAndPaused_issue96326369019436638851077/external-files/AnkiDroid/collection.media/
D/AbstractFlashcardViewer: fillFlashcard()
D/AbstractFlashcardViewer: base url = file:///var/folders/ym/nqynp93d4j74dpzw20sq4wq00000gn/T/robolectric-AbstractFlashcardViewerTest_noAutomaticAnswerAfterRenderProcessGoneAndPaused_issue96326369019436638851077/external-files/AnkiDroid/collection.media/
D/AbstractFlashcardViewer$LinkDetectingGestureDetector: Removing pending touch events for gestures
I/SoundPlayer: playing sounds for ANSWER
V/AbstractFlashcardViewer: onSoundGroupCompleted
I/AutomaticAnswer: Automatically showing question in 5000ms
I/AutomaticAnswer: stop: automatically show question
D/RobolectricTest: Calling destroy on controller com.ichi2.anki.AbstractFlashcardViewerTest$NonAbstractFlashcardViewer@700a863c
I/SoundPlayer: stopping sounds
D/SoundTagPlayer: Releasing sounds and abandoning audio focus
D/SoundTagPlayer: Abandoning audio focus
I/AnkiDroidApp$onCreate: NonAbstractFlashcardViewer::onDestroy
I/CollectionHelper: closeCollection: RobolectricTest: End
I/Collection: Collection closed

messages after flipping card
java.lang.AssertionError: messages after flipping card
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:26)
	at com.ichi2.anki.AbstractFlashcardViewerTest$noAutomaticAnswerAfterRenderProcessGoneAndPaused_issue9632$1.invokeSuspend(AbstractFlashcardViewerTest.kt:246)
	at com.ichi2.anki.AbstractFlashcardViewerTest$noAutomaticAnswerAfterRenderProcessGoneAndPaused_issue9632$1.invoke(AbstractFlashcardViewerTest.kt)
	at com.ichi2.anki.AbstractFlashcardViewerTest$noAutomaticAnswerAfterRenderProcessGoneAndPaused_issue9632$1.invoke(AbstractFlashcardViewerTest.kt)
	at com.ichi2.testutils.TestClass$runTest$1.invokeSuspend(TestClass.kt:163)
	at com.ichi2.testutils.TestClass$runTest$1.invoke(TestClass.kt)
	at com.ichi2.testutils.TestClass$runTest$1.invoke(TestClass.kt)
	at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTest$2$1$1.invokeSuspend(TestBuilders.kt:314)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
	at kotlinx.coroutines.test.TestDispatcher.processEvent$kotlinx_coroutines_test(TestDispatcher.kt:28)
	at kotlinx.coroutines.test.TestCoroutineScheduler.tryRunNextTaskUnless$kotlinx_coroutines_test(TestCoroutineScheduler.kt:103)
	at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTest$2$1$workRunner$1.invokeSuspend(TestBuilders.kt:320)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:280)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at kotlinx.coroutines.test.TestBuildersJvmKt.createTestResult(TestBuildersJvm.kt:13)
	at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTest-8Mi8wO0(TestBuilders.kt:308)
	at kotlinx.coroutines.test.TestBuildersKt.runTest-8Mi8wO0(Unknown Source)
	at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTest-8Mi8wO0(TestBuilders.kt:166)
	at kotlinx.coroutines.test.TestBuildersKt.runTest-8Mi8wO0(Unknown Source)
	at com.ichi2.testutils.TestClass.runTest(TestClass.kt:161)
	at com.ichi2.testutils.TestClass.runTest$default(TestClass.kt:155)
	at com.ichi2.anki.AbstractFlashcardViewerTest.noAutomaticAnswerAfterRenderProcessGoneAndPaused_issue9632(AbstractFlashcardViewerTest.kt:241)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:588)
	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$2(SandboxTestRunner.java:290)
	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:101)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
	Suppressed: org.robolectric.android.internal.AndroidTestEnvironment$UnExecutedRunnablesException: Main looper has queued unexecuted runnables. This might be the cause of the test failure. You might need a shadowOf(Looper.getMainLooper()).idle() call.
```

</details>

**Comment by mikehardy** [comment]
> I was offered financial sponsorship for TTS Support, but don't know if this is still going ahead. Let me know how you'd like to handle this

Do the work, get paid, IMHO. You're doing an amazing job fuzz-busting through *all* of the sound + tts stuff and I cannot believe for a second that whatever the arrangement it is it actually covers opportunity cost. I'll be reviewing + merging regardless so my time shouldn't be considered on a first pass, and on a second pass that goes against OpenCollective anyway. So passes on all sorts of criteria objectively if I've got my neutral arbitrator hat on and for me personally at least goes in to "heck yeah, that's fantastic" territory.

**Comment by mikehardy** [comment]
Not important at all: noticed on a URL that we were over 15000 and went to see what actually ended up as 15000, here we are! winner winner chicken dinner

**Review by mikehardy — APPROVED** [review]
That fifth commit though 😆 
I know you like the thorough reviews, and I do too - I read it...I don't see anything though. If I read it a few more times I might see something? Or just vote to merge it and let follow-ups happen once usage happens. Choosing the latter

**Comment by mikehardy** [comment]
took a conflict though, inevitably with all the refactor in flight

**Comment by david-allison** [comment]
> That fifth commit though 😆 I know you like the thorough reviews, and I do too - I read it...I don't see anything though. If I read it a few more times I might see something? Or just vote to merge it and let follow-ups happen once usage happens. Choosing the former

I tried 😆, there's not /much/ that I can do here: it's a rewrite, incremental commits would be additional risk + time for very little gain. Better to fix forward here + add tests if there are issues

**Review by BrayanDSO — APPROVED** [review]
Looks pretty good. I specially like the `SoundTagPlayer` split and the shift to coroutines.

It worked well here, besides the ReadText.kt functionality (like it reading anki:play:q) which isn't part of this PR.

---
I left comments to follow the rewrite wagon. Completely optional

File: AnkiDroid/src/main/java/com/ichi2/anki/AndroidTtsPlayer.kt [hunk 1]
```diff
@@ -215,3 +217,13 @@ class AndroidTtsError(@Suppress("unused") val errorCode: TtsErrorCode) : TtsPlay
             failure(TtsErrorCode.fromErrorCode(errorCode))
     }
 }
+
+fun TtsPlayer.TtsError.localizedErrorMessage(context: Context): String =
+    if (this is AndroidTtsError) {
+        // TODO: Do we want a human readable string here as well - snackbar has limited room
+        // but developerString is currently not translated as it returns
+        // developerString: ERROR_NETWORK_TIMEOUT, so "Audio error (ERROR_NETWORK_TIMEOUT)"
```
[the remark this answers is not in the document]
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/AndroidTtsPlayer.kt:225]: I prefer magic strings: easier to Google

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt [hunk 2]
```diff
@@ -1475,34 +1437,14 @@ abstract class AbstractFlashcardViewer :
...
-            // leaving the card (such as when edited)
-            mSoundPlayer.resetSounds()
-            mAnswerSoundsAdded = false
-            mSoundPlayer.addSounds(content.getSoundTags(Side.FRONT), SingleSoundSide.QUESTION)
-            if (automaticAnswer.isEnabled() && !mAnswerSoundsAdded && mCardSoundConfig!!.autoplay) {
-                addAnswerSounds { content.getSoundTags(Side.BACK) }
-            }
+        runBlocking {
+            soundPlayer.loadCardSounds(currentCard!!, if (displayAnswer) Side.BACK else Side.FRONT)
         }
```
[the remark this answers is not in the document]
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt:1448]: TODO it is

**Comment by david-allison** [comment]
Even better, thanks for the follow-ups

**Comment by BrayanDSO** [comment]
To be merged after CI is fixed and strings are synced

**Comment by BrayanDSO** [comment]
And for the money question, claim it and buy something you like 😉. If you hadn't bought the Anki socks already, that would be my suggestion

**Comment by david-allison** [comment]
Blocked on https://github.com/ankidroid/Anki-Android/labels/Strings

**Comment by mikehardy** [comment]
This was all ready to go and took a tiny conflict from #15061 - it was absolutely trivial to fix since I'd reviewed that one as well and I want to keep the train rolling so I've rebased+fixed it locally. Will push after it passes local testing then slam this one in

**Comment by mikehardy** [comment]
This was all ready to go and took a tiny conflict from #15064  - it was absolutely trivial to fix since I'd reviewed that one as well and I want to keep the train rolling so I've rebased+fixed it locally. Will push after it passes local testing then slam this one in

(if you look closely, you'll notice it's a second one from a second PR 😆 )

**Pull request merged**